### ISSUE TITLE
questions about training diffusion model

### ISSUE THREAD

**Issue by xu1px** [JuliaWolleb/diffusion-anomaly#2]
Special thanks for your outstanding work, I now want to train the diffusion model and I have successfully started training, but I don't know how many epoches are needed to train the diffusion model in total, and could you tell me how long does it take you to train the diffusion model?

**Comment by JuliaWolleb** [comment]
Hi
In the settings here, the model will be saved in all 10,000 steps (iterations). For my application, around 80,000 iterations were enough (with a batch size of 10) . This took around 1.5 days in my case. 
However, this depends on your dataset and the image-to-image translation task you want to solve. I would recommend to check the results (i.e., run the sampling process) from time to time on the saved models, to decide when the output images look good enough for you.

**Comment by xu1px** [comment]
> Hi In the settings here, the model will be saved in all 10,000 steps (iterations). For my application, around 80,000 iterations were enough (with a batch size of 10) . This took around 1.5 days in my case. However, this depends on your dataset and the image-to-image translation task you want to solve. I would recommend to check the results (i.e., run the sampling process) from time to time on the saved models, to decide when the output images look good enough for you.

Thank you very much for your answer, I have now successfully reproduced the results of your experiment.

**Comment by 0Godness** [comment]
> 

hello, how do you get the AUROC and dice score using this code. It seems no metric code in the current version. Would you mind sharing the code that you successfully reproduced the results? Thanks!

**Comment by xu1px** [comment]
Sorry for interrupting again, I found a discrepancy between the dice score in my reproduced results and your results, and I think this may be a problem with the dataset. According to the paper, the middle 49 slices are selected, but the total number of both training and test sets cannot be divided by 49, so I would like to ask if there are other filtering operations on the dataset in the experiment, and could you tell me which samples you have selected from brats2020 so that I can better reproduce the results in the paper.

**Comment by xu1px** [comment]
> > 
> 
> hello, how do you get the AUROC and dice score using this code. It seems no metric code in the current version. Would you mind sharing the code that you successfully reproduced the results? Thanks!



> > 
> 
> hello, how do you get the AUROC and dice score using this code. It seems no metric code in the current version. Would you mind sharing the code that you successfully reproduced the results? Thanks!

Apologies for the late reply, I just used the otsu thresholds from the opencv library as described in the original paper, and did not add additional operations.

**Comment by xu1px** [comment]
> > 
> 
> hello, how do you get the AUROC and dice score using this code. It seems no metric code in the current version. Would you mind sharing the code that you successfully reproduced the results? Thanks!

Apologies for the late reply, I just used the otsu thresholds from the opencv library as described in the original paper, and did not add additional operations.

**Comment by JuliaWolleb** [comment]
Hi
I checked the numbers for training and testing and you are right, there is a typo: In the training set, we have 5,598 healthy slices, and 10,670 (instead of 10,607)  diseased slices. The sum is can then be divided by 59.
On the test set, we ignored the slices that are right on the edge of the tumor. That is, we discarded slices where the tumor is smaller than 10 pixels.
For the evaluation metrics, I added a file in the folder /scripts with a broad overview of the functions I used. But it is correct, for the dice score I thresholded the anomaly map with the Otsu method.

**Comment by kazusa597** [comment]
thanks for sharing the code very much! I want to train the model on my own dataset, but the test result didn't guide the disease data to the healthy data.  I trained the classifier model using both disease data and healthy data, is it right?

**Comment by kazusa597** [comment]
Also I didn't use Visdom,  Does this affect model training？

> thanks for sharing the code very much! I want to train the model on my own dataset, but the test result didn't guide the disease data to the healthy data. I trained the classifier model using both disease data and healthy data, is it right?

**Comment by JuliaWolleb** [comment]
Hi
Not using Visdom does not affect model training. this is just a visualization tool.
Yes, you need the classifier as well as the diffusion model to train on both healthy and diseased data. What is the issue you face? You could check whether your classifier works properly if you set `maxt=0` in line 55 of the file _classifier_train.py_, and then check the classification accuracy. `maxt=1000` means that it is trained on noisy images with t in {1,...,1000}.

**Comment by kazusa597** [comment]
> 

![903output](https://user-images.githubusercontent.com/116068629/198959810-835545b9-ccd9-41f8-a909-1a3b65d07956.jpg)
![903input](https://user-images.githubusercontent.com/116068629/198959804-024ad7e0-04c8-447d-8ece-6d05d8ef4463.jpg)

**Comment by kazusa597** [comment]
I used ground penetrating radar (GPR) data，but the output is similar to the input. I think maybe these data are not suitable. 
![903input](https://user-images.githubusercontent.com/116068629/198959977-317d2bbe-1397-4002-9184-39bfcbe7a08a.jpg)
![903output](https://user-images.githubusercontent.com/116068629/198959981-a9ce8233-24ae-4b97-893d-1654455e8213.jpg)

**Comment by JuliaWolleb** [comment]
Eh, so what was the goal of your task? I think this data should be suitable, if you change from one class to the other? did you try to tune L and s? How do your gradient maps look like during sampling?

**Comment by kazusa597** [comment]
em, I try to split the different signals. Yes, i trained the classifier as well.  maybe i should change the gradient scale s. Thank you again！
![图片1](https://user-images.githubusercontent.com/116068629/199020523-f0f92ee1-09a1-4e58-b95e-3dd7e44e5880.png)
![图片2](https://user-images.githubusercontent.com/116068629/199020538-fcaceeb8-6724-4cf4-bd71-39eabb72eb90.png)

**Comment by JuliaWolleb** [comment]
I see. You need two groups of images for your training, one only containing clean signals and the other one containing defected signals. Then the classifier should provide gradient maps that remove the defected signal.

**Comment by tungnthust** [comment]
> > Hi In the settings here, the model will be saved in all 10,000 steps (iterations). For my application, around 80,000 iterations were enough (with a batch size of 10) . This took around 1.5 days in my case. However, this depends on your dataset and the image-to-image translation task you want to solve. I would recommend to check the results (i.e., run the sampling process) from time to time on the saved models, to decide when the output images look good enough for you.
> 
> Thank you very much for your answer, I have now successfully reproduced the results of your experiment.

Can you share your training settings to reproduce the results in the paper ? I have tried but the results are not good. 
I'm confused about some settings below:
- How do you normalize data to value between 0 and 1 ?
- Did you use all default hyper-parameters in source code, or is there any change ?
- Did you train with timesteps t from 0 to T = 1000, or just train with t from 0 to L=500

**Comment by Junyiliu0** [comment]
> Hi
> In the settings here, the model will be saved in all 10,000 steps (iterations). For my application, around 80,000 iterations were enough (with a batch size of 10) . This took around 1.5 days in my case.
> However, this depends on your dataset and the image-to-image translation task you want to solve. I would recommend to check the results (i.e., run the sampling process) from time to time on the saved models, to decide when the output images look good enough for you.

Hi
Can you tell me how iterations did you use for training the chexpert_dataset? I found maybe the batch size of the cheapest_dataset is 1. And how long cost?
Besides,
I found there will be two .pt file during training.
<img width="278" alt="image" src="https://github.com/JuliaWolleb/diffusion-anomaly/assets/100564143/8ea7fc6f-27b6-4b85-b23e-b49ddebe55a0">
Which one should I choose during testing?  What is meaning of ema?
Hope your reply!
Thanks!

**Comment by JuliaWolleb** [comment]
Hi
As pointed out in the paper, we trained the  DDPM model for 50,000 iterations. This took roughly 2 days.
The meaning of ema is "exponential moving average". I picked the other one (brats2update080000.pt) for inference.
